File: src/crypto.py
import logging
from pathlib import Path

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import padding
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

from book import BookFormat, BookInfo, FileKind
from metadata import extract_title, sanitize_filename

logger = logging.getLogger("ridi")


def _is_valid_output(fmt: BookFormat, data: bytes) -> bool:
    match fmt:
        case BookFormat.EPUB:
            return data.startswith((b"PK\x03\x04", b"PK\x05\x06", b"PK\x07\x08"))
        case BookFormat.PDF:
            return data.startswith(b"%PDF")


def decrypt_key(book_info: BookInfo, device_id: str, debug: bool = False) -> bytes:
    """Extract AES session key from the encrypted .dat file.

    Layout: first 16 bytes = IV, rest = AES-128-CBC ciphertext (PKCS7).
    The 16-byte session key sits at byte offset 68..84 of the plaintext.
    """
    data_path = book_info.get_file(FileKind.DATA)
    if not data_path.exists():
        raise FileNotFoundError(f"Missing data file: {data_path}")

    data = data_path.read_bytes()

    if debug:
        logger.debug("Data file: %s (%d bytes)", data_path, len(data))

    # first 16 bytes of device_id → AES key; first 16 bytes of .dat → IV
    key = device_id.encode("utf-8")[:16]
    iv = data[:16]

    cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
    decryptor = cipher.decryptor()
    decrypted = decryptor.update(data[16:]) + decryptor.finalize()

    unpadder = padding.PKCS7(128).unpadder()
    plaintext = unpadder.update(decrypted) + unpadder.finalize()

    if len(plaintext) < 84:
        raise ValueError(f".dat plaintext too short: {len(plaintext)} bytes")

    # session key: 16 ASCII bytes at positions 68..84
    session_key = plaintext[68:84].decode("utf-8", errors="ignore").encode("utf-8")
    if len(session_key) != 16:
        raise ValueError("Invalid session key length")

    if debug:
        logger.debug("Session key: %s", session_key.hex())

    return session_key


def decrypt_book(book_info: BookInfo, key: bytes, debug: bool = False) -> bytes:
    """Decrypt the book file using the session key (AES-128-CBC + PKCS7).

    If the file already looks like a valid EPUB/PDF, it is returned as-is.
    """
    book_file_path = book_info.get_file(FileKind.BOOK)

    if not book_file_path.exists():
        raise FileNotFoundError(f"Book file not found: {book_file_path}")

    raw = book_file_path.read_bytes()

    if debug:
        logger.debug("Book file: %s (%d bytes)", book_file_path, len(raw))

    if _is_valid_output(book_info.format, raw):
        if debug:
            logger.debug("File already valid; returning as-is")
        return raw

    if len(raw) < 16:
        raise ValueError("Book file too small to contain IV")

    iv = raw[:16]
    cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
    decryptor = cipher.decryptor()
    decrypted = decryptor.update(raw[16:]) + decryptor.finalize()

    unpadder = padding.PKCS7(128).unpadder()
    return unpadder.update(decrypted) + unpadder.finalize()


def decrypt_and_save(
    book_info: BookInfo,
    device_id: str,
    debug: bool = False,
    output_dir: Path | None = None,
) -> Path:
    key = decrypt_key(book_info, device_id, debug)
    content = decrypt_book(book_info, key, debug)

    title = extract_title(book_info.format, content)
    if title:
        out_name = f"{sanitize_filename(title)}.{book_info.format.extension()}"
    else:
        out_name = book_info.file_name(FileKind.BOOK)

    out_dir = output_dir or Path.cwd()
    target = out_dir / out_name

    if target.exists():
        stem, suffix = target.stem, target.suffix
        i = 1
        while target.exists() and i < 1000:
            target = out_dir / f"{stem} ({i}){suffix}"
            i += 1

    target.write_bytes(content)
    if debug:
        logger.debug("Wrote output: %s", target)
    return target


def decrypt_with_progress(
    book_info: BookInfo,
    device_id: str,
    debug: bool = False,
    output_dir: Path | None = None,
) -> bool:
    file_name = book_info.file_name(FileKind.BOOK)

    print(f'\r\u28ff Decrypting "{file_name}"', end="", flush=True)

    try:
        decrypt_and_save(book_info, device_id, debug, output_dir)
        print(f'\r\u28ff Decrypting "{file_name}" \u2714\ufe0e')
        return True
    except Exception as e:
        print(f'\r\u28ff Decrypting "{file_name}" \u2718')
        if debug:
            logger.error("Error decrypting %s: %s", book_info.id, e)
        return False

File: src/crypto.ts
import { createDecipheriv } from "crypto";
import { readFileSync, writeFileSync, existsSync } from "fs";
import { join } from "path";
import { BookInfo, FileKind, bookFormatExtension, BookFormat } from "./book.js";
import { extractTitle, sanitizeFilename } from "./metadata.js";

function isValidOutput(fmt: BookFormat, data: Buffer): boolean {
  if (fmt === BookFormat.EPUB) {
    return (
      data.subarray(0, 4).equals(Buffer.from([0x50, 0x4b, 0x03, 0x04])) ||
      data.subarray(0, 4).equals(Buffer.from([0x50, 0x4b, 0x05, 0x06])) ||
      data.subarray(0, 4).equals(Buffer.from([0x50, 0x4b, 0x07, 0x08]))
    );
  } else if (fmt === BookFormat.PDF) {
    return data.subarray(0, 4).equals(Buffer.from("%PDF"));
  }
  return false;
}

function removePKCS7Padding(data: Buffer): Buffer {
  const paddingLength = data[data.length - 1];
  if (paddingLength > 16 || paddingLength === 0) {
    throw new Error("Invalid PKCS7 padding");
  }
  for (let i = 1; i <= paddingLength; i++) {
    if (data[data.length - i] !== paddingLength) {
      throw new Error("Invalid PKCS7 padding");
    }
  }
  return data.subarray(0, data.length - paddingLength);
}

export function decryptKey(bookInfo: BookInfo, deviceId: string, debug: boolean = false): Buffer {
  const dataPath = bookInfo.getFile(FileKind.DATA);
  if (!existsSync(dataPath)) {
    throw new Error(`Missing data file: ${dataPath}`);
  }

  const data = readFileSync(dataPath);

  if (debug) {
    console.log(`Data file: ${dataPath} (${data.length} bytes)`);
  }

  const key = Buffer.from(deviceId.substring(0, 16), "utf-8");
  const iv = data.subarray(0, 16);

  const decipher = createDecipheriv("aes-128-cbc", key, iv);
  decipher.setAutoPadding(false);
  const decrypted = Buffer.concat([decipher.update(data.subarray(16)), decipher.final()]);

  const plaintext = removePKCS7Padding(decrypted);

  if (plaintext.length < 84) {
    throw new Error(`.dat plaintext too short: ${plaintext.length} bytes`);
  }

  const sessionKey = Buffer.from(plaintext.subarray(68, 84).toString("utf-8", 0, 16), "utf-8");
  if (sessionKey.length !== 16) {
    throw new Error("Invalid session key length");
  }

  if (debug) {
    console.log(`Session key: ${sessionKey.toString("hex")}`);
  }

  return sessionKey;
}

export function decryptBook(bookInfo: BookInfo, key: Buffer, debug: boolean = false): Buffer {
  const bookFilePath = bookInfo.getFile(FileKind.BOOK);

  if (!existsSync(bookFilePath)) {
    throw new Error(`Book file not found: ${bookFilePath}`);
  }

  const raw = readFileSync(bookFilePath);

  if (debug) {
    console.log(`Book file: ${bookFilePath} (${raw.length} bytes)`);
  }

  if (isValidOutput(bookInfo.format, raw)) {
    if (debug) {
      console.log("File already valid; returning as-is");
    }
    return raw;
  }

  if (raw.length < 16) {
    throw new Error("Book file too small to contain IV");
  }

  const iv = raw.subarray(0, 16);
  const decipher = createDecipheriv("aes-128-cbc", key, iv);
  decipher.setAutoPadding(false);
  const decrypted = Buffer.concat([decipher.update(raw.subarray(16)), decipher.final()]);

  return removePKCS7Padding(decrypted);
}

export async function decryptAndSave(
  bookInfo: BookInfo,
  deviceId: string,
  debug: boolean = false,
  outputDir?: string
): Promise<string> {
  const key = decryptKey(bookInfo, deviceId, debug);
  const content = decryptBook(bookInfo, key, debug);

  const title = await extractTitle(bookInfo.format, content);
  let outName: string;
  if (title) {
    outName = `${sanitizeFilename(title)}.${bookFormatExtension(bookInfo.format)}`;
  } else {
    outName = bookInfo.fileName(FileKind.BOOK);
  }

  const outDir = outputDir || process.cwd();
  let target = join(outDir, outName);

  if (existsSync(target)) {
    const lastDot = outName.lastIndexOf(".");
    const stem = lastDot > 0 ? outName.substring(0, lastDot) : outName;
    const suffix = lastDot > 0 ? outName.substring(lastDot) : "";
    let i = 1;
    while (existsSync(target) && i < 1000) {
      target = join(outDir, `${stem} (${i})${suffix}`);
      i++;
    }
  }

  writeFileSync(target, content);
  if (debug) {
    console.log(`Wrote output: ${target}`);
  }
  return target;
}

export async function decryptWithProgress(
  bookInfo: BookInfo,
  deviceId: string,
  debug: boolean = false,
  outputDir?: string
): Promise<boolean> {
  const fileName = bookInfo.fileName(FileKind.BOOK);

  process.stdout.write(`\r⣿ Decrypting "${fileName}"`);

  try {
    await decryptAndSave(bookInfo, deviceId, debug, outputDir);
    console.log(`\r⣿ Decrypting "${fileName}" ✔︎`);
    return true;
  } catch (e) {
    console.log(`\r⣿ Decrypting "${fileName}" ✘`);
    if (debug) {
      console.error(`Error decrypting ${bookInfo.id}: ${e}`);
    }
    return false;
  }
}
